File: dynamicProgramming/bestTimeToBuyAndSellStock.java
package dynamicProgramming;

// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
// Solution : https://www.youtube.com/watch?v=4YjEHmw1MX0


public class bestTimeToBuyAndSellStock {
    public static void main(String[] args) {
        
    }
}


class Solution {

    
    // TC : O(N^2
    // SC : O(1)

    //Brute Force
    public int maxProfit(int[] prices) {

        int maxProfit = 0;

        for (int i = 0 ; i< prices.length ; i++){
            for (int j = i ; j < prices.length; j++){
                maxProfit = Math.max(maxProfit,prices[j]- prices[i]);
            }
        }
        return maxProfit;
    }

    // TC : O(N)
    // SC : O(1)

    // Runtime: 2 ms, faster than 78.41% of Java online submissions for Best Time to Buy and Sell Stock.
    // Memory Usage: 52.1 MB, less than 69.52% of Java online submissions for Best Time to Buy and Sell Stock.

    public int maxProfitApproach2(int[] prices) {

        int leastStockSoFar= prices[0];
        int maxProfit = 0;

        for (int i: prices){
            leastStockSoFar = Math.min(i, leastStockSoFar);
            maxProfit = Math.max(maxProfit, i-leastStockSoFar);
        }
        return maxProfit;
    }
}

File: dynamicProgramming/bestTimeToBuyAndSellStock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
# Solution : https://www.youtube.com/watch?v=4YjEHmw1MX0


class Solution(object):
    
    
    #TC: O(N^2)
    # SC: O(1)
    
    # BruteForce : TLE
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        maximum = 0
        
        for i in range(len(prices)-1):
            for j in range(i+1, len(prices)):
                maximum = max(maximum, prices[j]-prices[i])
                    
        return maximum
            
    
    
    # TC: O(N)
    # SC: O(1)
    
    # Efficient
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        leastPriceOfStockSoFar = prices[0]
        profitMadeToday, overallProfit =0,  0
        
        for i in range(1,len(prices)):
            leastPriceOfStockSoFar = min(leastPriceOfStockSoFar, prices[i])
            profitMadeToday= prices[i]-leastPriceOfStockSoFar
            overallProfit= max(overallProfit, profitMadeToday)
            
        return overallProfit
    
print(Solution().maxProfit([7,1,2,5,6,4]))